### DISCUSSION TITLE
Is there something like emacs input-method mechanism in SWM?

### DISCUSSION THREAD

**Discussion by spiderbit** [stumpwm/stumpwm#1140] [Ideas]
I use in emacs german-postfix input-method that allows me to create german umlauts with ae, ue, oe and sz.

If I type ae followed with another e it creates only "ae" but if after the ae follows any other letter it generates the german letter ä.

Now I found the function: "define-remapped-keys" in the stumpwm documentation and wonder if I only be able to use this keyboards with modifiers because for stumpwm commands I can use:
(kbd "u") as a shortcut in a keymap. So could you create a keymap per 2 letter combination or is there anything like input methods in stumpwm?

**Comment by equwal** [comment]
You could change the X keymap with setxkbmap. You'd need a keymap that does want you want though.

**Comment by spiderbit** [comment]
I am not sure that is possible, but the problem is there are different contexts, so I would need such translation in input-modes if you will or text-modes, but it can be problematic in command modes. Often I have bound the letter "a" to functions. But I guess that makes the thing application specific, because stumpwm can't differentiate if I am in a text-field in a application at least not without some heavy communication between the application and the window-manager.

So I guess it's a applicationspecific problem. Or I must edit text fields with emacs with Atomicchrome / Ghosttext, maybe that get's me something close at least.

**Comment by equwal** [comment]
That sure is asking a lot from your window manager. FWIW I just use a compose key.

**Comment by dmb2** [comment]
@spiderbit In principle this sounds like it should work.  If I'm understanding your question, conceptually you want to:
1. Define multiple top-level maps, one for each keypair that will eventually become a special character
2. Define sub-keys that make up a pair and when pressed send the accented key to the client.

I haven't used redefine-keys myself, but it looks like its oriented to be application specific.  If you got something kind of hacked together, it would make a nice module for contrib :).

One problem I'm not sure how you'll solve is the case when you want a literal ae instead of the accented case.  Maybe the command that is triggered when you type 'ae' will wait for another input and only output the accent if you type an additional e, and aeX where X is whatever key is pressed otherwise.

This could make for a neat project, but its definitely outside the scope of what StumpWM is trying to support.

**Comment by spiderbit** [comment]
Yeah again the problem is that this should only happen in insert-mode, because if I press a "a" in command mode, it should not wait if I press a "e" afterwards but just do whatever I put on the "a" key as command.
So in emacs I use xah-fly-key a modal mode like in Vim, and there I activate this german input-mode in edit-mode and deactivate it in command mode with a hook.

**Comment by spiderbit** [comment]
Is there a way to ask a X application if a textfield is currently activated? Or get some sort of hook, well either per polling or by other means.

**Comment by dmb2** [comment]
I'm not sure... I honestly don't know X well enough to make a definitive comment.  Since this isn't a bug or issue, I'm going to make it a discussion.
